### DISCUSSION TITLE
Storage .upload call - how to set up the file to send in the call - Dart

### DISCUSSION THREAD

**Discussion by ptadev55** [supabase/supabase#28168] [Questions] [answered]
Has anyone had success using the .upload call from Dart? Specifically, I seem to be having an issue with the datatype required for the file to be passed in correctly. 

In my example below, can anyone tell me what datatype/setup is required for the file parm in the .upload ? At one time I remember seeing in the docs, that you could pass in a "File" type or a "Uint8list". I don't see that anymore, but regardless, I've tried both and I can't get either to work. If anyone has any insight or a working example, I'd love to hear it/see it!

```
// let's assume for this example that I:
/      started with a image Url (in the imageUrl variable)
//     did a GET to get the image file
//     placed the response in a Uint8List
final response = await http.get(Uri.parse(imageUrl)); 
Uint8List imageData = response.bodyBytes;
// Million $ question: What do I do to get it into a variable that the .upload will accept??

final supaclient = SupaFlow.client;

final String uploadedUrl = await supaclient.storage.from('content').upload(
    'images/myfiletoupload.png',
   ???, // place the right type of variable here (in place of the ???)
    fileOptions: const FileOptions(cacheControl: '3600', upsert: true),
);
```

**Comment by GaryAustin1** [comment, marked as answer]
I don't know Flutter/Dart but the docs show a different upload method for Uin8List...
![image](https://github.com/user-attachments/assets/75dca184-e1ff-42c2-81e2-2910641625e6)

**Reply by ptadev55** [reply]
Thanks @GaryAustin1 .. I didn't notice that "Upload file on web" button when I was looking at the examples. Hopefully this works in my Flutter app too; I just tried this and I'm as step further :) now I have 403 permission errors to the bucket I'm going at.  Once I get that resolved, I will share my example code here.

**Reply by ptadev55** [reply]
Quick update. It's working like a charm. I set my RLS permissions for Auth, Anon at the bucket level (that's what Flutterflow needs), but I could have gotten more granular (by folder if need be).. all the permission details for storage are here: https://supabase.com/docs/guides/storage/security/access-control, and the call @GaryAustin1 pointed me to is defined here: https://supabase.com/docs/reference/dart/storage-from-upload (note that the .upload was the one I was having the trouble with, but by switching to the "Upload file on Web" instructions, and using the .uploadBinary, it accepted the Uint8list as I expected. 

See code example below (note in my example, this is a Flutterflow custom action where I had taken a incoming imageUrl, resized the image, then uploaded the resized image with an appropriate name. I may optimize further, but you'll get the idea below):

```
Future<String?> uploadResizedImage(String imageUrl, int width) async {
  try {
    // Get filename for upload call
    Uri incomingUri = Uri.parse(imageUrl);
    String fileName = incomingUri.pathSegments.last;
    // find the .png extension
    int dotIndex = fileName.lastIndexOf('.');
    // Create a new filename by appending _{width} to the fileName, and then .png
    String resizedFileName = fileName.substring(0, dotIndex) +
        '_' +
        width.toString() +
        fileName.substring(dotIndex);
    print('uploadResizedImage: resizedFileName: ${resizedFileName}');

    // Fetch image data from URL
    final response = await http.get(Uri.parse(imageUrl));
    if (response.statusCode != 200) {
      print('uploadResizedImage: Failed to load image from URL: $imageUrl');
      return null;
    }

    // Decode image data
    Uint8List imageData = response.bodyBytes;
    img.Image? image = img.decodeImage(imageData);
    if (image == null) {
      print('uploadResizedImage: Failed to decode image from loaded data');
      return null;
    }

    // Resize the image
    print('Original image size: ${image.length} bytes');
    img.Image resizedImage = img.copyResize(image, width: width);
    print(
        'uploadResizedImage: Resized image size: ${resizedImage.length} bytes');

    // Encode the resized image to uploadable bytes
    Uint8List resizedImageBytes = Uint8List.fromList(img.encodePng(resizedImage));
    print('uploadResizedImage: Image encoded to PNG format');

    // supabase example with my parameters in it
    final supaclient = SupaFlow.client;

    // Upload the resized image to Supabase Storage
    final String resizedUrl = await supaclient.storage
        .from('{put_your_bucket_name_here}')
        .uploadBinary(
          '{put_your_folder_here_if_applicable}/${resizedFileName}',
          resizedImageBytes,
          fileOptions: const FileOptions(cacheControl: '3600', upsert: true),
        );

    print('uploadResizedImage: upload complete, file: ${resizedFileName}');
    // Return the URL of the resized image
    return resizedUrl;
  } catch (e) {
    print('uploadResizedImage: Error: $e');
    return null;
    // show error and return null
  }
}
```
